### PR TITLE
[Android] Allow to use system fonts 

### LINKED ISSUE #11422: Unable to load font 'monospace'
### Description

I'm using the 'monospace' font in my xamarin forms app.
When migrating to maui, it looks like it does not work anymore

### Steps to Reproduce

1. Create a maui app for android
2. Create a label using the 'monospace' font
3. Run the app

`<Label FontFamily="monospace" Text="Hello world" />`

### Link to public reproduction project repository

...

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

...

### Did you find any workaround?

_No response_

### Relevant log output

```shell
[Microsoft.Maui.FontManager] [Warning] Unable to load font 'monospace' from assets.Java.Lang.RuntimeException: Font asset not found monospace
[Microsoft.Maui.FontManager]    at Java.Interop.JniEnvironment.StaticMethods.CallStaticObjectMethod(JniObjectReference type, JniMethodInfo method, JniArgumentValue* args) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniEnvironment.g.cs:line 12890
[Microsoft.Maui.FontManager]    at Java.Interop.JniPeerMembers.JniStaticMethods.InvokeObjectMethod(String encodedMember, JniArgumentValue* parameters) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniStaticMethods.cs:line 95
[Microsoft.Maui.FontManager]    at Android.Graphics.Typeface.CreateFromAsset(AssetManager mgr, String path) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net6.0/android-33/mcw/Android.Graphics.Typeface.cs:line 613
[Microsoft.Maui.FontManager]    at Microsoft.Maui.FontManager.LoadTypefaceFromAsset(String fontfamily, Boolean warning) in D:\a\_work\1\s\src\Core\src\Fonts\FontManager.Android.cs:line 122
[Microsoft.Maui.FontManager]   --- End of managed Java.Lang.RuntimeException stack trace ---
[Microsoft.Maui.FontManager] java.lang.RuntimeException: Font asset not found monospace
[Microsoft.Maui.FontManager] 	at android.graphics.Typeface.createFromAsset(Typeface.java:1044)
[Microsoft.Maui.FontManager] 	at crc6452ffdc5b34af3a0f.NavigationViewFragment.n_onCreateView(Native Method)
[Microsoft.Maui.FontManager] 	at crc6452ffdc5b34af3a0f.NavigationViewFragment.onCreateView(NavigationViewFragment.java:39)
[Microsoft.Maui.FontManager] 	at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2995)
[Microsoft.Maui.FontManager] 	at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:523)
[Microsoft.Maui.FontManager] 	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:261)
[Microsoft.Maui.FontManager] 	at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:1840)
[Microsoft.Maui.FontManager] 	at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1764)
[Microsoft.Maui.FontManager] 	at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1701)
[Microsoft.Maui.FontManager] 	at androidx.fragment.app.FragmentManager$4.run(FragmentManager.java:488)
[Microsoft.Maui.FontManager] 	at android.os.Handler.handleCallback(Handler.java:942)
[Microsoft.Maui.FontManager] 	at android.os.Handler.dispatchMessage(Handler.java:99)
[Microsoft.Maui.FontManager] 	at android.os.Looper.loopOnce(Looper.java:201)
[Microsoft.Maui.FontManager] 	at android.os.Looper.loop(Looper.java:288)
[Microsoft.Maui.FontManager] 	at android.app.ActivityThread.main(ActivityThread.java:7898)
[Microsoft.Maui.FontManager] 	at java.lang.reflect.Method.invoke(Native Method)
[Microsoft.Maui.FontManager] 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
[Microsoft.Maui.FontManager] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change

Allow to use Android system fonts directly.

```
<VerticalStackLayout VerticalOptions="Center">
    <Label
        Text="Hello world" />
    <Label
        FontFamily="monospace"
        Text="Hello world" />
    <Label
        FontFamily="serif"
        Text="Hello world" />
</VerticalStackLayout>
```

Screenshot_1683643513
image

### Issues Fixed

Fixes #11422

### CODE REVIEW

File: src/Core/src/Fonts/FontManager.Android.cs [hunk 1]
```diff
@@ -130,6 +135,21 @@ public FontSize GetFontSize(Font font, float defaultFontSize = 0)
 			return null;
 		}
 
+		Typeface? LoadDefaultTypeface(string fontfamily)
+		{
+			switch (fontfamily.ToLowerInvariant())
+			{
+				case "monospace":
+					return Typeface.Monospace;
+				case "sansserif":
```
Comment (mattleibow) [inline src/Core/src/Fonts/FontManager.Android.cs:144]: Should we add some "sans-serif" - and any CSS equivalents? What would you do in Native Android / CSS to get these fonts? I think `sans-serif` at least?

File: src/Core/src/Fonts/FontManager.Android.cs [hunk 2]
```diff
@@ -130,6 +135,21 @@ public FontSize GetFontSize(Font font, float defaultFontSize = 0)
...
+					return Typeface.SansSerif;
+				case "serif":
+					return Typeface.Serif;
+				default:
+					return null;
+			}
+		}
+
 		Typeface? CreateTypeface((string? fontFamilyName, FontWeight weight, bool italic) fontData)
```
Comment (mattleibow) [inline src/Core/src/Fonts/FontManager.Android.cs:153]: I wonder if this method is the correct place? `CreateTypeface`

The `LoadTypefaceFromAsset` is very specific so maybe moving this to check before - in the caller?

Comment (jsuarezruiz) [reply src/Core/src/Fonts/FontManager.Android.cs:150]: Good point, I have reviewed the code and I think the changes are now in the right place.

**Review by jfversluis — APPROVED** [review]

**Review by mattleibow — APPROVED** [review]

**Comment by jonathanpeppers** [comment]
@jsuarezruiz these changes cause an exception to be thrown several times on startup in an app I've been testing:

```
06-19 15:31:19.834  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium' from assets.
06-19 15:31:19.840  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium.ttf' from assets.
06-19 15:31:19.840  5706  5706 I DOTNET  : Unable to load font 'Fonts/sans-serif-medium.ttf' from assets.
06-19 15:31:19.841  5706  5706 I DOTNET  : Unable to load font 'fonts/sans-serif-medium.ttf' from assets.
06-19 15:31:19.841  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium.otf' from assets.
06-19 15:31:19.842  5706  5706 I DOTNET  : Unable to load font 'Fonts/sans-serif-medium.otf' from assets.
06-19 15:31:19.842  5706  5706 I DOTNET  : Unable to load font 'fonts/sans-serif-medium.otf' from assets.
```

https://github.com/dotnet/maui/blob/503b5344c40e49b234c0c7a654881250ecdb550e/src/Core/src/Fonts/FontManager.Android.cs#L127-L128

On a Pixel 5, `dotnet-trace` shows a decent amount of time doing this:

```
64.98ms  microsoft.maui!Microsoft.Maui.Handlers.LabelHandler.MapFont(Microsoft.Maui.Handlers.ILabelHandler,Microsoft.Maui.ILabel)
```

I open the Shell flyout in this app to see these:

https://github.com/supershopping/ShellFlyoutLagging

Is there something going wrong here?

**Comment by jsuarezruiz** [comment]
> @jsuarezruiz these changes cause an exception to be thrown several times on startup in an app I've been testing:
> 
> ```
> 06-19 15:31:19.834  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium' from assets.
> 06-19 15:31:19.840  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium.ttf' from assets.
> 06-19 15:31:19.840  5706  5706 I DOTNET  : Unable to load font 'Fonts/sans-serif-medium.ttf' from assets.
> 06-19 15:31:19.841  5706  5706 I DOTNET  : Unable to load font 'fonts/sans-serif-medium.ttf' from assets.
> 06-19 15:31:19.841  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium.otf' from assets.
> 06-19 15:31:19.842  5706  5706 I DOTNET  : Unable to load font 'Fonts/sans-serif-medium.otf' from assets.
> 06-19 15:31:19.842  5706  5706 I DOTNET  : Unable to load font 'fonts/sans-serif-medium.otf' from assets.
> ```
> 
> https://github.com/dotnet/maui/blob/503b5344c40e49b234c0c7a654881250ecdb550e/src/Core/src/Fonts/FontManager.Android.cs#L127-L128
> 
> (NOTE: I changed the above to `Console.WriteLine()` and fixed the missing `$` locally)
> 
> On a Pixel 5, `dotnet-trace` shows a decent amount of time doing this:
> 
> ```
> 64.98ms microsoft.maui!Microsoft.Maui.Handlers.LabelHandler.MapFont(Microsoft.Maui.Handlers.ILabelHandler,Microsoft.Maui.ILabel)
> 38.92ms microsoft.maui!Microsoft.Maui.FontManager.LoadTypefaceFromAsset(string,bool)
> ```
> 
> I open the Shell flyout in this app to see these:
> 
> https://github.com/supershopping/ShellFlyoutLagging
> 
> Is there something going wrong here?

I think you found this one: https://github.com/dotnet/maui/issues/13239
Searching the possible origin, could be:

https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/Platform/Android/BottomNavigationViewUtils.cs#L218
https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellItemRenderer.cs#L248

**Comment by jonathanpeppers** [comment]
Ok, I think we should add a couple more cases: https://github.com/dotnet/maui/pull/15759

It seems to help performance to avoid these lookups.

**Comment by ewerspej** [comment]
Is there an equivalent for iOS? I'm trying to use an iOS-specific font in a project that has been converted from XF to MAUI, like this:

```xml
<Style x:Key="bodySmallSemibold" TargetType="Label">
    <Setter Property="FontFamily">
        <Setter.Value>
            <OnPlatform x:TypeArguments="x:String">
                <On Platform="iOS">.SFUI-Semibold</On>
                <On Platform="Android">sans-serif-medium</On>
            </OnPlatform>
        </Setter.Value>
    </Setter>
</Style>
```

But the font is not used. Should I open a ticket for this or is this already being worked on/known?

**Pull request merged**